# Crime Prediction

In [56]:
import pandas
import numpy
from sklearn.tree import DecisionTreeClassifier, export_graphviz
from sklearn import metrics
from sklearn.ensemble import RandomForestClassifier
from sklearn import tree, svm, linear_model, preprocessing
from sklearn.pipeline import Pipeline
import io
from scipy import misc
import pydotplus
import graphviz
import matplotlib.pyplot as plt
from IPython.display import Image
from sklearn.model_selection import cross_val_score, GridSearchCV
from sklearn.naive_bayes import GaussianNB

In [57]:
oriDF = pandas.read_csv("communities-crime-clean.csv")
oriDF

state        communityname  fold  population  householdsize  \
0         1        Alabastercity     7        0.01           0.61   
1         1    AlexanderCitycity    10        0.01           0.41   
2         1         Annistoncity     3        0.03           0.34   
3         1           Athenscity     8        0.01           0.38   
4         1           Auburncity     1        0.04           0.37   
5         1         Bessemercity     6        0.04           0.44   
6         1       Birminghamcity     2        0.41           0.37   
7         1          Cullmancity     1        0.01           0.30   
8         1           Daphnecity     7        0.00           0.39   
9         1          Decaturcity    10        0.06           0.39   
10        1           Dothancity     4        0.07           0.41   
11        1       Enterprisecity     2        0.02           0.43   
12        1          Eufaulacity     5        0.01           0.46   
13        1        Fairfieldcity     1        0.00           0.45   
14        1         Florencecity     3        0.04           0.33   
15        1        FortPaynecity     2        0.00           0.35   
16        1          Gadsdencity     9        0.05           0.32   
17        1        Hartsellecity     4        0.00           0.43   
18        1         Homewoodcity     5        0.02           0.23   
19        1           Hoovercity    10        0.05           0.35   
20        1         Hueytowncity     4        0.01           0.44   
21        1       Huntsvillecity     1        0.24           0.37   
22        1     Jacksonvillecity     7        0.00           0.54   
23        1           Jaspercity     8        0.01           0.37   
24        1          Madisoncity     5        0.01           0.36   
25        1           Mobilecity     8        0.30           0.41   
26        1       Montgomerycity     2        0.28           0.45   
27        1    MountainBrookcity     8        0.02           0.37   
28        1        Northportcity     2        0.01           0.44   
29        1          Opelikacity     6        0.02           0.45   
...     ...                  ...   ...         ...            ...   
1963     55        NewBerlincity     4        0.04           0.55   
1964     55         OakCreekcity     6        0.02           0.49   
1965     55       Oconomowoccity     6        0.00           0.42   
1966     55         Onalaskacity     4        0.00           0.44   
1967     55          Oshkoshcity     7        0.07           0.42   
1968     55           Racinecity    10        0.12           0.44   
1969     55       RiverFallscity     3        0.00           0.67   
1970     55        Sheboygancity     8        0.06           0.37   
1971     55     Shorewoodvillage     6        0.01           0.18   
1972     55   SouthMilwaukeecity     9        0.02           0.39   
1973     55     StevensPointcity     4        0.02           0.51   
1974     55       SunPrairiecity     2        0.01           0.48   
1975     55         Superiorcity     8        0.03           0.34   
1976     55        TwoRiverscity     6        0.00           0.37   
1977     55        Watertowncity     2        0.00           0.52   
1978     55         Waukeshacity     6        0.08           0.45   
1979     55           Wausaucity     7        0.04           0.37   
1980     55        Wauwatosacity     9        0.06           0.35   
1981     55        WestAlliscity     6        0.09           0.29   
1982     55         WestBendcity     3        0.02           0.49   
1983     55  WhitefishBayvillage     9        0.01           0.42   
1984     55       Whitewatercity     6        0.00           0.65   
1985     55  WisconsinRapidscity     8        0.01           0.32   
1986     56         Cheyennecity     4        0.06           0.34   
1987     56         Evanstoncity     6        0.00           0.65   
1988     56         Gillettecity     9        0.01           0.53   
1989

## 1: Decision Tree Classification

### a:

In [58]:
oriDF['highCrime'] = (oriDF['ViolentCrimesPerPop'] > 0.1)
oriDF[['ViolentCrimesPerPop', 'highCrime']]

ViolentCrimesPerPop highCrime
0                    0.06     False
1                    0.14      True
2                    1.00      True
3                    0.23      True
4                    0.15      True
5                    1.00      True
6                    1.00      True
7                    0.16      True
8                    0.05     False
9                    0.22      True
10                   0.52      True
11                   0.41      True
12                   0.28      True
13                   0.80      True
14                   0.25      True
15                   0.03     False
16                   0.90      True
17                   0.01     False
18                   0.15      True
19                   0.09     False
20                   0.22      True
21                   0.55      True
22                   0.19      True
23                   0.53      True
24                   0.08     False
25                   0.44      True
26                   0.35      True
27                   0.02     False
28                   0.35      True
29                   1.00      True
...                   ...       ...
1963                 0.05     False
1964                 0.05     False
1965                 0.07     False
1966                 0.02     False
1967                 0.06     False
1968                 0.34      True
1969                 0.04     False
1970                 0.06     False
1971                 0.07     False
1972                 0.05     False
1973                 0.10     False
1974                 0.03     False
1975                 0.11      True
1976                 0.02     False
1977                 0.04     False
1978                 0.06     False
1979                 0.02     False
1980                 0.07     False
1981                 0.10     False
1982                 0.05     False
1983                 0.01     False
1984                 0.06     False
1985                 0.11      True
1986                 0.08     False
1987                 0.11      True
1988                 0.15      True
1989                 0.06     False
1990                 0.12      True
1991                 0.27      True
1992                 0.06     False

[1993 rows x 2 columns]

In [59]:
noofT = oriDF['highCrime'].value_counts()[0]
noofF = oriDF['highCrime'].value_counts()[1]
totalVals = noofT + noofF
print("Percent of \"True\" instances are: "+ str(format(float((noofT*100)/totalVals), '.2f')) + " %")
print("Percent of \"False\" instances are: "+ str(format(float((noofF*100)/totalVals), '.2f')) + " %")

Percent of "True" instances are: 37.28 %
Percent of "False" instances are: 62.72 %


### b: Decision Tree Classifier

In [60]:
allCols = list(oriDF)

exclu = ['state', 'communityname', 'fold', 'ViolentCrimesPerPop', 'highCrime']
inclu = [item for item in allCols if item not in exclu]

modDF = oriDF[inclu]
modDF.data = modDF.copy()
modDF.target = oriDF['highCrime']

In [61]:
model = DecisionTreeClassifier()
model.fit(modDF.data, modDF.target)
print(model)

expected = modDF.target
predicted = model.predict(modDF.data)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_split=1e-07, min_samples_leaf=1,
            min_samples_split=2, min_weight_fraction_leaf=0.0,
            presort=False, random_state=None, splitter='best')


In [62]:
print(metrics.classification_report(expected, predicted))
print(metrics.confusion_matrix(expected, predicted))

             precision    recall  f1-score   support

      False       1.00      1.00      1.00       743
       True       1.00      1.00      1.00      1250

avg / total       1.00      1.00      1.00      1993

[[ 743    0]
 [   0 1250]]


In [63]:
important_features = pandas.DataFrame(model.feature_importances_, columns = ["Importance"], index = modDF.data.columns).sort_values("Importance", ascending=False)
print("Importance of Features(Top 10)")
important_features[:10]

Importance of Features(Top 10)


Importance
PctKids2Par         0.358652
racePctWhite        0.087891
racePctHisp         0.048585
PctLess9thGrade     0.020673
PctEmplManu         0.020141
PctSpeakEnglOnly    0.014980
PctImmigRec8        0.013952
PctHousOwnOcc       0.013686
PctNotHSGrad        0.013448
MedRent             0.013290

In [64]:
def show_tree(decisionTree, file_path):
    dotfile = io.StringIO()
    export_graphviz(decisionTree, out_file=dotfile)
    pydotplus.graph_from_dot_data(dotfile.getvalue()).write_png(file_path)
    i = misc.imread(file_path)
    plt.imshow(i)

# To use it
show_tree(model, 'test.png')

<img src="test.png">

Important Features Summary:

In the Decision tree above we can see that the highest impportance is given to the feature "PctKids2Par". The decision tree may predict states that the crime rate is high where percent of kids having both parents is low, for this we can come to the conclusion that the single parents might be inclined towards comitting crime to support their children all by him/her self. 

The other defining attirbutes also include percent of different races living in that area and we can also see education of the people is also an importanrt facter. Despite these attributes which doesnt make any meaning like amount of people working in manifacturing still we might want to consider the percent of people under and above poverty line because it makes sense.

### b: Decision Tree Classifier with 10 fold cross validation

In [65]:
model1 = DecisionTreeClassifier()
print(model1)

def APR(model):
    accuracy = cross_val_score(model, modDF.data, modDF.target, cv = 10, scoring='accuracy')
    precision = cross_val_score(model, modDF.data, modDF.target, cv = 10, scoring='precision')
    recall = cross_val_score(model, modDF.data, modDF.target, cv = 10, scoring='recall')
    print("Accuracy: %0.2f (+/- %0.2f)" % (accuracy.mean(), accuracy.std() * 2))
    print("Precision: %0.2f (+/- %0.2f)" % (precision.mean(), precision.std() * 2))
    print("Recall: %0.2f (+/- %0.2f)" % (recall.mean(), recall.std() * 2))

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_split=1e-07, min_samples_leaf=1,
            min_samples_split=2, min_weight_fraction_leaf=0.0,
            presort=False, random_state=None, splitter='best')


#### i) Accuracy, Percision and Recall for Decision Tree Classifier with 10 fold cross validation

In [66]:
APR(model1)

Accuracy: 0.72 (+/- 0.09)
Precision: 0.79 (+/- 0.05)
Recall: 0.77 (+/- 0.18)


#### ii) Alalysis on Accuracy, Percision and Recall

In the initial decision tree classification we can see that all Accuracy, Percision and Recall were 100 % because the model built was overfitted to the training set but once we applied cross validation the data was split into train and test and the model was buit on that and now the Accuracy, Percision and Recall for the new model went down more than by 20 % and the new model is not overfitted towrds the training set and thus we can assume that the new model will perform better when encountered with new datapoints.

## 2: Linear Classification

### a. Naive Bayes Classifier 

#### i) Naive Bayes Classification with 10 fold Cross Validation

In [67]:
gnbModel = GaussianNB()

print(gnbModel)

GaussianNB(priors=None)


In [68]:
APR(gnbModel)

Accuracy: 0.76 (+/- 0.11)
Precision: 0.91 (+/- 0.12)
Recall: 0.69 (+/- 0.18)


#### ii) Top 10 Features according to their Predictiveness

In [69]:
def Predictiveness(cname):
    tVals = numpy.array(oriDF.loc[oriDF.highCrime == True, cname])
    fVals = numpy.array(oriDF.loc[oriDF.highCrime == False, cname])
    nume = abs(tVals.mean() - fVals.mean())
    deno = tVals.std() + fVals.std()
    val = float('%.4f' % (nume/deno))
    return val

def PredictiveFeatures(df):
    tupList = []
    colNames = list(df.columns.values)
    for cname in colNames:
        tup = (cname, Predictiveness(cname))
        tupList.append(tup)
    return tupList

labels = ['Feature', 'Predictiveness']
PredictiveFeaturesList = pandas.DataFrame.from_records(PredictiveFeatures(modDF), columns=labels).sort_values("Predictiveness", ascending=False)
PredictiveFeaturesList[:10]

Feature  Predictiveness
44       PctKids2Par          0.8097
43        PctFam2Par          0.7455
3       racePctWhite          0.7352
50          PctIlleg          0.7093
40      FemalePctDiv          0.6940
41       TotalPctDiv          0.6746
45  PctYoungKids2Par          0.6650
15        pctWInvInc          0.6611
46       PctTeen2Par          0.6429
38    MalePctDivorce          0.6169

In the above result we can see the the feature "PctKids2Par" is counted as the most predictive among other feature and this is same the most important feature from the decision tree classification. This phenomenon makes sense because the formule on which the predictiveness is directly proportional to the difference in the means of features with their respective classes this is because the mean represents the data spread of that feature for that respective class hence the higher the difference between means the easier to differentiate between the classes.

#### iii) Comparison between the Feature Predictiveness and Important Features from Decision Tree Classification

Form the above results we can see that their are many features differences between "Features with high Predictiveness" and "Important Features" but we can see that there are two common features('PctKids2Par', 'racePctWhite') in the top three features from both lists by which we can assume that these features have a big take in the decision making because they showed up in both ways of finding important features.

### b) linear Support Vector Classification

#### i) Linear Support Vector Classification with 10 fold Cross Validation

In [70]:
linearModel = svm.LinearSVC()
print(linearModel)

LinearSVC(C=1.0, class_weight=None, dual=True, fit_intercept=True,
     intercept_scaling=1, loss='squared_hinge', max_iter=1000,
     multi_class='ovr', penalty='l2', random_state=None, tol=0.0001,
     verbose=0)


In [71]:
APR(linearModel)

Accuracy: 0.80 (+/- 0.11)
Precision: 0.85 (+/- 0.10)
Recall: 0.83 (+/- 0.23)


In [72]:
linearModel.fit(modDF.data, modDF.target)

LinearSVC(C=1.0, class_weight=None, dual=True, fit_intercept=True,
     intercept_scaling=1, loss='squared_hinge', max_iter=1000,
     multi_class='ovr', penalty='l2', random_state=None, tol=0.0001,
     verbose=0)

#### ii) Feature Weights - Top 10

In [73]:
feature_weights = pandas.DataFrame(abs(linearModel.coef_[0]), columns = ["Weight"], index = modDF.data.columns).sort_values("Weight", ascending=False)
feature_weights[:10]

Weight
pctWInvInc        1.888486
PersPerOccupHous  1.755127
racePctWhite      1.500219
PctKids2Par       1.190329
RentHighQ         1.066885
MalePctDivorce    1.065696
NumUnderPov       1.051548
NumStreet         1.019155
PctOccupMgmtProf  1.014671
population        1.002301

In the above result, we can see attributes having the absolute weights. The highest weight corresponds to the attribute “PersPerOccupHous” which tells us the number of persons living per household and this could affect the crime rate in that area. There are some other attributes which also contribute to the crime rate like races and poverty line but there are some attributes with high weights but doesn’t quit affect the crime rate like “pctWInvInc: percentage of households with investment / rent income in 1989” but we should not just rule this out because they might indirectly affect the crime rate. From this analysis, we can see that the weights of the attributes can give the importance of that attribute in decision making and this is because the coefficients represent the vector orthogonal to the linear hyper plane which separates the data and the higher the coefficient’s absolute value the further it is away from the plane and the further it is away gives more info on classification of the data points. Thus, these weights give us the most predictive features in the dataset.

#### iii) Comparison between the Feature Weights and Important Features from Decision Tree Classification

When we compare the above results with the top 10 important features from decision tree classification we can see that we have some new attributes which we have high weights/importance like ‘PersPerOccupHous’, ‘pctWInvInc’, ‘NumStreet’, etc., these attributes did not show up in the previous but managed to rack up high weights and the thing to observe is that we have some attributes which are common from both list of features like ‘racePctWhite’ and ‘PctKids2Par’. These feature managed to get into the most influencing attributes lists from different classification techniques and this tells us that these attributes have a influencing vote on the deciding the classification of a data point.

## 3: Regression

### a) Linear Regression

In [74]:
linRegModel = linear_model.LinearRegression()
linRegModel.fit(modDF.data, oriDF['ViolentCrimesPerPop'])
lrmExpec = oriDF['ViolentCrimesPerPop'].copy()
lrmPredic = linRegModel.predict(modDF.data)

#### i) Mean Squared Error with 10 fold Cross Validation

In [75]:
mse = cross_val_score(linear_model.LinearRegression(), modDF.data, oriDF['ViolentCrimesPerPop'], cv=10, scoring='neg_mean_squared_error')
print("Mean Squared Error: %0.2f (+/- %0.2f)" % (mse.mean(), mse.std() * 2))

Mean Squared Error: -0.02 (+/- 0.01)


#### ii) Mean Squared Error

In [76]:
mse1 = metrics.mean_squared_error(lrmExpec, lrmPredic)
print("Mean Squared Error: %0.4f" % mse1)

Mean Squared Error: 0.0165


#### iii) 

##### Top 10 Features for predicting High Crime Rate

In [77]:
feature_weights_pVe = pandas.DataFrame(linRegModel.coef_, columns = ["Weight"], index = modDF.data.columns).sort_values("Weight", ascending=False)
feature_weights_pVe[:10]

Weight
PersPerOccupHous  0.635088
PctHousOwnOcc     0.568133
MalePctDivorce    0.458517
PctRecImmig8      0.432511
MedRent           0.372728
medFamInc         0.287979
PctEmploy         0.248474
MalePctNevMarr    0.226728
PctPersDenseHous  0.214353
OwnOccMedVal      0.212876

##### Top 10 Features for predicting Low Crime Rate

In [78]:
feature_weights_nVe = pandas.DataFrame(linRegModel.coef_, columns = ["Weight"], index = modDF.data.columns).sort_values("Weight", ascending=True)
feature_weights_nVe[:10]

Weight
PctPersOwnOccup    -0.675694
TotalPctDiv        -0.561924
whitePerCap        -0.351016
PctKids2Par        -0.322651
OwnOccLowQuart     -0.308170
numbUrban          -0.296443
PersPerRentOccHous -0.254572
RentLowQ           -0.234752
agePct12t29        -0.229218
PctRecImmig5       -0.218221

### b) Ridge Regression

In [79]:
#alpha=[10.0, 1.0, 0.1, 0.01, 0.001]
rModel = linear_model.Ridge()
rModel.fit(modDF.data, oriDF['ViolentCrimesPerPop'])
rExpec = oriDF['ViolentCrimesPerPop'].copy()
rPredic = rModel.predict(modDF.data)

#### i) Mean Squared Error with 10 fold Cross Validation

In [80]:
rMse = cross_val_score(rModel, modDF.data, oriDF['ViolentCrimesPerPop'], cv=10, scoring='neg_mean_squared_error')
print("Estimated Mean Squared Error: %0.2f (+/- %0.2f)" % (rMse.mean(), rMse.std() * 2))

Estimated Mean Squared Error: -0.02 (+/- 0.01)


#### ii) Mean Squared Error

In [81]:
rMse1 = metrics.mean_squared_error(rExpec, rPredic)
print("Mean Squared Error: %0.4f" % rMse1)

Mean Squared Error: 0.0168


#### iii) Best Alpha

In [82]:
rcvModel = linear_model.RidgeCV(alphas=[10.0, 1.0, 0.1, 0.01, 0.001])
rcvModel.fit(modDF.data, oriDF['ViolentCrimesPerPop'])
print("Best Alpha from [10.0, 1.0, 0.1, 0.01, 0.001] is: %0.1f" % rcvModel.alpha_)

Best Alpha from [10.0, 1.0, 0.1, 0.01, 0.001] is: 1.0


#### iv)

The MSE computed when we used Linear Regression is 0.0165 and when we used Ridge Regression then is 0.0168 we can observe that there is an increase in the squared error. When a model is overfitted to the training set then the squared error is closer towards 0 and 0 when the model is completely overfitted. So, in this scenario we can see that the MSE for Linear Regression is lower than Ridge thus we can assume that Linear Regression model is more overfitted than the Ridge model for this dataset. 

### c) Polynomial Regression

#### i) Mean Squared Error with 10 fold Cross Validation

In [83]:
pModel = Pipeline([('poly', preprocessing.PolynomialFeatures(degree=2)), ('linear', linear_model.LinearRegression())])
pModel.fit(modDF.data, oriDF['ViolentCrimesPerPop'])
pExpec = oriDF['ViolentCrimesPerPop'].copy()
pPredic = pModel.predict(modDF.data)

In [84]:
pMse = cross_val_score(pModel, modDF.data, oriDF['ViolentCrimesPerPop'], cv=10, scoring='neg_mean_squared_error')
print("Mean Squared Error: %0.2f (+/- %0.2f)" % (pMse.mean(), pMse.std() * 2))

Mean Squared Error: -0.13 (+/- 0.08)


#### ii) Mean Squared Error

In [85]:
pMse1 = metrics.mean_squared_error(pExpec, pPredic)
print("Mean Squared Error: %0.4f" % pMse1)
pMse1

Mean Squared Error: 0.0000


8.1944380154569338e-29

#### iii)

The linear regression will draw a linear plane to classify any new data points but as for the polynomial regression will form a polynomial plane which will try to fit to the plane to the data points this could lead to reduced sum of squares or even overfitting. When we have scored, the mean squared error for the dataset using the polynomial regression and we can see that the mean squared error went down very close to 0.0 when we compare with the linear regression and this can be because of the overfitting in the polynomial regression and we are trying to predict a numeric value rather than the True/False on which we have done in the previous experiments. But if applied the cross validation when the model is built we can see that the linear regression is being over fitted because the mean squared error is closer to zero but the cross-validation score for ploy is not that far from zero. From this analysis above we can consider linear regression to be a better fit because chance of overfitting is more in polynomial regression. 

## 4: Dirty Data

In [86]:
oriDF2 = pandas.read_csv("communities-crime-full.csv")
oriDF2 = oriDF2.replace('?', numpy.NaN)
oriDF2

state county community           communityname  fold  population  \
0         8    NaN       NaN            Lakewoodcity     1        0.19   
1        53    NaN       NaN             Tukwilacity     1        0.00   
2        24    NaN       NaN            Aberdeentown     1        0.00   
3        34      5     81440     Willingborotownship     1        0.04   
4        42     95      6096       Bethlehemtownship     1        0.01   
5         6    NaN       NaN       SouthPasadenacity     1        0.02   
6        44      7     41500             Lincolntown     1        0.01   
7         6    NaN       NaN               Selmacity     1        0.01   
8        21    NaN       NaN           Hendersoncity     1        0.03   
9        29    NaN       NaN             Claytoncity     1        0.01   
10        6    NaN       NaN            DalyCitycity     1        0.13   
11       36    NaN       NaN  RockvilleCentrevillage     1        0.02   
12       25     21     44105             Needhamtown     1        0.03   
13       55     87     30075          GrandChutetown     1        0.01   
14        6    NaN       NaN           DanaPointcity     1        0.04   
15       19    187     91370           FortDodgecity     1        0.03   
16       36      1      1000              Albanycity     1        0.15   
17       34     27     17650        Denvilletownship     1        0.01   
18       18    NaN       NaN          Valparaisocity     1        0.02   
19       42    129     66376       Rostravertownship     1        0.00   
20        6    NaN       NaN             Modestocity     1        0.25   
21       12     31       NaN        Jacksonvillecity     1        1.00   
22       41    NaN       NaN        KlamathFallscity     1        0.01   
23       19    193     93926           SiouxCitycity     1        0.11   
24        6    NaN       NaN              Delanocity     1        0.02   
25        8    NaN       NaN              Goldencity     1        0.00   
26        6    NaN       NaN             Gardenacity     1        0.06   
27       39     29     61798           Perrytownship     1        0.01   
28       54    NaN       NaN             Beckleycity     1        0.01   
29        9      7     22490         EastHamptontown     1        0.00   
...     ...    ...       ...                     ...   ...         ...   
1964     36    119     49121         MountVernoncity    10        0.09   
1965     18     97       NaN        Indianapoliscity    10        1.00   
1966      6    NaN       NaN         RedwoodCitycity    10        0.09   
1967     25     17     76135            Westfordtown    10        0.01   
1968     34     25     62430          RedBankborough    10        0.00   
1969      6    NaN       NaN        SanFranciscocity    10        1.00   
1970     45    NaN       NaN               Caycecity    10        0.00   
1971      1    NaN       NaN             Decaturcity    10        0.06   
1972      6    NaN       NaN            Redlandscity    10        0.08   
1973     39    NaN       NaN           Englewoodcity    10        0.00   
1974     37    NaN       NaN             Tarborotown    10        0.00   
1975     53    NaN       NaN            Kirklandcity    10        0.05   
1976     25     17      1605           Arlingtontown    10        0.06   
1977     47    NaN       NaN           Maryvillecity    10        0.01   
1978      9      1      4720              Betheltown    10        0.01   
1979     34     39     61530              Rahwaycity    10        0.02   
1980     25      5     69170             Tauntoncity    10        0.06   
1981      9      9     35650              Hamdentown    10        0.07   
1982     42    133     73168    SpringGardentownship    10        0.00   
1983     13    NaN       NaN          ForestParkcity    10        0.01   
1984     39     95     48342              Maumeecity    10        0.01   
1985      1    NaN       NaN       AlexanderCitycity    10        0.01   
1986      

In [87]:
oriDF2['highCrime'] = (oriDF2['ViolentCrimesPerPop'] > 0.1)
oriDF2[['ViolentCrimesPerPop', 'highCrime']]

ViolentCrimesPerPop highCrime
0                    0.20      True
1                    0.67      True
2                    0.43      True
3                    0.12      True
4                    0.03     False
5                    0.14      True
6                    0.03     False
7                    0.55      True
8                    0.53      True
9                    0.15      True
10                   0.24      True
11                   0.08     False
12                   0.06     False
13                   0.09     False
14                   0.21      True
15                   0.30      True
16                   0.49      True
17                   0.07     False
18                   0.15      True
19                   0.03     False
20                   0.34      True
21                   0.69      True
22                   0.21      True
23                   0.63      True
24                   0.31      True
25                   0.12      True
26                   0.84      True
27                   0.10     False
28                   0.49      True
29                   0.02     False
...                   ...       ...
1964                 0.60      True
1965                 0.69      True
1966                 0.30      True
1967                 0.04     False
1968                 0.19      True
1969                 0.75      True
1970                 0.53      True
1971                 0.22      True
1972                 0.30      True
1973                 0.04     False
1974                 0.52      True
1975                 0.09     False
1976                 0.08     False
1977                 0.25      True
1978                 0.04     False
1979                 0.16      True
1980                 0.28      True
1981                 0.07     False
1982                 0.03     False
1983                 0.56      True
1984                 0.14      True
1985                 0.14      True
1986                 0.02     False
1987                 0.04     False
1988                 0.19      True
1989                 0.09     False
1990                 0.45      True
1991                 0.23      True
1992                 0.19      True
1993                 0.48      True

[1994 rows x 2 columns]

In [88]:
noofT2 = oriDF2['highCrime'].value_counts()[0]
noofF2 = oriDF2['highCrime'].value_counts()[1]
totalVals2 = noofT2 + noofF2
print("Percent of \"True\" instances are: "+ str(format(float((noofT2*100)/totalVals2), '.2f')) + " %")
print("Percent of \"False\" instances are: "+ str(format(float((noofF2*100)/totalVals2), '.2f')) + " %")

Percent of "True" instances are: 37.26 %
Percent of "False" instances are: 62.74 %


In [89]:
allCols2 = list(oriDF)

exclu2 = ['state', 'communityname', 'fold', 'ViolentCrimesPerPop', 'highCrime']
inclu2 = [item for item in allCols2 if item not in exclu2]

modDF2 = oriDF2[inclu2]
modDF2.data = modDF2.copy()
modDF2.target = oriDF2['highCrime']

values = modDF2.data.values
# fill missing values with mean column values
imputer = preprocessing.Imputer()
transformed_modDF2_data = imputer.fit_transform(values)







In [90]:
model2 = DecisionTreeClassifier()
model2.fit(transformed_modDF2_data, modDF2.target)
print(model2)

expected2 = modDF2.target
predicted2 = model2.predict(transformed_modDF2_data)

print(metrics.classification_report(expected2, predicted2))
print(metrics.confusion_matrix(expected2, predicted2))

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_split=1e-07, min_samples_leaf=1,
            min_samples_split=2, min_weight_fraction_leaf=0.0,
            presort=False, random_state=None, splitter='best')
             precision    recall  f1-score   support

      False       1.00      1.00      1.00       743
       True       1.00      1.00      1.00      1251

avg / total       1.00      1.00      1.00      1994

[[ 743    0]
 [   0 1251]]


In [91]:
important_features2 = pandas.DataFrame(model2.feature_importances_, columns = ["Importance"], index = modDF2.data.columns).sort_values("Importance", ascending=False)
print("Importance of Features(Top 10)")
important_features2[:10]


Importance of Features(Top 10)


Importance
PctKids2Par        0.361370
racePctWhite       0.088089
racePctHisp        0.045107
PctEmplManu        0.025833
PctLess9thGrade    0.020817
PctHousOwnOcc      0.014161
PctEmploy          0.013360
HousVacant         0.011952
OtherPerCap        0.011732
householdsize      0.011265

In [92]:
accuracy = cross_val_score(model2,transformed_modDF2_data, modDF2.target, cv = 10, scoring='accuracy')
precision = cross_val_score(model2, transformed_modDF2_data, modDF2.target, cv = 10, scoring='precision')
recall = cross_val_score(model2, transformed_modDF2_data, modDF2.target, cv = 10, scoring='recall')
print("Accuracy: %0.2f (+/- %0.2f)" % (accuracy.mean(), accuracy.std() * 2))
print("Precision: %0.2f (+/- %0.2f)" % (precision.mean(), precision.std() * 2))
print("Recall: %0.2f (+/- %0.2f)" % (recall.mean(), recall.std() * 2))

Accuracy: 0.75 (+/- 0.05)
Precision: 0.81 (+/- 0.05)
Recall: 0.79 (+/- 0.05)


### a) Analysis:

On comparison with the previous scores of Accuracy, Precision and Recall of the clean dataset and the scores from above dirty dataset we can see that the scores have improved. This is because when handling the missing we used imputing to fill out the missing values and due to this we have more information to classify the tree unlike the clean dataset where eliminated the missing values. Thus we have obtained better Accuracy, Precision and Recall scores.

## 5:(6 a) Extra Credit

### 5. a) 

### i) Non-Linear Kernel for SVM

#### Clena Data

In [93]:
svm3Model = svm.SVC(degree = 4)
svm3Model.fit(modDF.data, modDF.target)

SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=4, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)

In [94]:
APR(svm3Model)

Accuracy: 0.82 (+/- 0.10)
Precision: 0.86 (+/- 0.11)
Recall: 0.86 (+/- 0.20)


#### Full Data

In [95]:
accuracy11 = cross_val_score(svm3Model,transformed_modDF2_data, modDF2.target, cv = 10, scoring='accuracy')
precision11 = cross_val_score(svm3Model, transformed_modDF2_data, modDF2.target, cv = 10, scoring='precision')
recall11 = cross_val_score(svm3Model, transformed_modDF2_data, modDF2.target, cv = 10, scoring='recall')
print("Accuracy: %0.2f (+/- %0.2f)" % (accuracy11.mean(), accuracy11.std() * 2))
print("Precision: %0.2f (+/- %0.2f)" % (precision11.mean(), precision11.std() * 2))
print("Recall: %0.2f (+/- %0.2f)" % (recall11.mean(), recall11.std() * 2))

Accuracy: 0.83 (+/- 0.02)
Precision: 0.86 (+/- 0.03)
Recall: 0.87 (+/- 0.04)


### Random Forest

#### Clena Data

In [96]:
rfModel = RandomForestClassifier()
rfModel

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_split=1e-07, min_samples_leaf=1,
            min_samples_split=2, min_weight_fraction_leaf=0.0,
            n_estimators=10, n_jobs=1, oob_score=False, random_state=None,
            verbose=0, warm_start=False)

In [97]:
APR(rfModel)

Accuracy: 0.79 (+/- 0.11)
Precision: 0.84 (+/- 0.09)
Recall: 0.81 (+/- 0.23)


#### Full Data

In [98]:
accuracy12 = cross_val_score(svm3Model,transformed_modDF2_data, modDF2.target, cv = 10, scoring='accuracy')
precision12 = cross_val_score(svm3Model, transformed_modDF2_data, modDF2.target, cv = 10, scoring='precision')
recall12 = cross_val_score(svm3Model, transformed_modDF2_data, modDF2.target, cv = 10, scoring='recall')
print("Accuracy: %0.2f (+/- %0.2f)" % (accuracy12.mean(), accuracy11.std() * 2))
print("Precision: %0.2f (+/- %0.2f)" % (precision12.mean(), precision11.std() * 2))
print("Recall: %0.2f (+/- %0.2f)" % (recall12.mean(), recall11.std() * 2))

Accuracy: 0.83 (+/- 0.02)
Precision: 0.86 (+/- 0.03)
Recall: 0.87 (+/- 0.04)


### ii) Result Comparisions

When we compare the Accuracy, Percision adn Recall for a non lenear model SVM and a Random Forest and we can see that the scores are better when we used non lenear model.  

### iii) Predictive Features

In [99]:
rfModel.fit(modDF.data, modDF.target)
important_features3 = pandas.DataFrame(rfModel.feature_importances_, columns = ["Importance"], index = modDF2.data.columns).sort_values("Importance", ascending=False)
print("Importance of Features(Top 10)")
important_features3[:10]

Importance of Features(Top 10)


Importance
NumIlleg            0.099234
PctKids2Par         0.065218
PctPersDenseHous    0.061595
PctTeen2Par         0.060323
PctFam2Par          0.044441
racepctblack        0.030988
pctWInvInc          0.026689
PctPersOwnOccup     0.026129
racePctHisp         0.025919
FemalePctDiv        0.022035

We can see that from the various lists of top predictive features using various models which we have a set of common attributes which show up in various lists, some of these are ‘PctFam2Par’, ‘PctKids2Par’, ‘racePctWhite’ and these attributes being able to predict the final predictions seems to be very reliable.  